### PR TITLE
use cached build files from app when making a build

### PR DESCRIPTION
**Pull request by snopoke**
## Summary
Part 2 of https://dimagi-dev.atlassian.net/browse/USH-1225

Prevent re-generating app build files when creating a build. The build files are currently generated during validation which is the first step when creating a build. The function where these files are generated is memoized but when we make a build we call the same function again on the copy of the app so the cache isn't used.

This PR uses the build files from the main app and just attaches them to the copy when creating a build.

## Product Description
Improve performance of creating a build.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [ ] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
App build is already covered by test.

### QA Plan
None

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/app_manager/tests/test_shadow_forms.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def test_resource_not_added(self, *args):
         expected = """
             <partial>
               <xform>
-                <resource id="advanced_module_form_0">
+                <resource descriptor="Form: (Module advanced_module module) - advanced_module form 0" id="advanced_module_form_0">
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/app_manager/tests/test_shadow_forms.py:64]: This changed because previously the app did not have a build spec set.

**Review by snopoke — COMMENTED** [review]
@esoergel @orangejenny what do you think of this approach?

The changes to tests are all related to either the lack of build spec in some test apps or else cached / memoize fields that weren't being cleared.

**Comment by orangejenny** [comment]
Are there any issues with the two docs having different ids? I'm thinking of the URLs in the build profile, won't those now reference the canonical app instead of the new build?

If that isn't a problem, I think this approach is great.

**Comment by snopoke** [comment]
> Are there any issues with the two docs having different ids? I'm thinking of the URLs in the build profile, won't those now reference the canonical app instead of the new build?

You are probably right - if that's the case then it's going to be way harder to reuse the build artifacts. I'll do some more digging.

**Comment by esoergel** [comment]
Would it be possible to make the copy first, then do the validation on that?  This might mean that the copy will need to be in-memory only and not saved until the end, or perhaps saved with a flag denoting that it's an in-progress build or something.

Another way of thinking of it would be: rather than validating that a build will work, then building, just try building and accept that it might fail.

**Comment by snopoke** [comment]
> Would it be possible to make the copy first, then do the validation on that? This might mean that the copy will need to be in-memory only and not saved until the end, or perhaps saved with a flag denoting that it's an in-progress build or something.
> 
> Another way of thinking of it would be: rather than validating that a build will work, then building, just try building and accept that it might fail.

yes, that's my next approach

**Comment by orangejenny** [comment]
> This might mean that the copy will need to be in-memory only and not saved until the end, or perhaps saved with a flag denoting that it's an in-progress build or something.

Maybe created with doc type `Application-Deleted` and then change to `Application` as the last step.

**Comment by snopoke** [comment]
New approach: https://github.com/dimagi/commcare-hq/pull/30405

**Pull request closed without merging**